### PR TITLE
introduce common abstraction for CDC via debezium

### PR DESCRIPTION
**Pull request by subodh1810**
Issue : https://github.com/airbytehq/airbyte/issues/3934

The PR introduces abstraction for debezium implementation. Across our DB connectors (mysql, postgres and sql server) we are using debezium for handling CDC. A lot of debezium implementation is common across multiple connectors and as a result we can introduce an abstraction to re-use code so that its easy to implement CDC across more DB connectors without having too much duplicate code

## Pre-merge Checklist
Expand the checklist which is relevant for this PR. 

<details><summary> <strong> Connector checklist </strong> </summary>
<p>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] Secrets are annotated with `airbyte_secret` in the connector's spec
- [ ] Credentials added to Github CI if needed and not already present. [instructions for injecting secrets into CI](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#using-credentials-in-ci). 
- [ ] Unit & integration tests added as appropriate (and are passing)
    * Community members: please provide proof of this succeeding locally e.g: screenshot or copy-paste acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] `/test connector=connectors/<name>` command as documented [here](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#updating-an-existing-connector) is passing. 
    * Community members can skip this, Airbyters will run this for you. 
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] `README.md`
    - [ ] `docs/SUMMARY.md` if it's a new connector
    - [ ] Created or updated reference docs in `docs/integrations/<source or destination>/<name>`.
    - [ ] Changelog in the appropriate page in `docs/integrations/...`. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md` contains a reference to the new connector
    - [ ] Build status added to [build page](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/builds.md)
- [ ] Build is successful
- [ ] Connector version bumped like described [here](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#updating-a-connector)
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#updating-a-connector)
- [ ] No major blockers
- [ ] PR merged into master branch
- [ ] Follow up tickets have been created
- [ ] Associated tickets have been closed & stakeholders notified
</p>
</details>

<details><summary> <strong> Connector Generator checklist </strong> </summary>
<p>
   
- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed.
</p>
</details>

### CODE REVIEW

File: airbyte-integrations/connectors/source-debezium/build.gradle [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    implementation 'io.debezium:debezium-connector-postgres:1.4.2.Final'
+
+    testFixturesImplementation project(':airbyte-db')
+    testFixturesImplementation project(':airbyte-integrations:bases:base-java')
+
+    testFixturesImplementation 'org.junit.jupiter:junit-jupiter-engine:5.4.2'
+    testFixturesImplementation 'org.junit.jupiter:junit-jupiter-api:5.4.2'
+    testFixturesImplementation 'org.junit.jupiter:junit-jupiter-params:5.4.2'
+
+}
```
Comment (davinchia) [inline airbyte-integrations/connectors/source-debezium/build.gradle:23]: nit: newline

File: airbyte-integrations/connectors/source-debezium/src/main/java/io/airbyte/integrations/source/debezium/DebeziumEventUtils.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        .withData(data);
+
+    return new AirbyteMessage()
+        .withType(AirbyteMessage.Type.RECORD)
+        .withRecord(airbyteRecordMessage);
+  }
+
+  // warning mutates input args.
+  private static JsonNode formatDebeziumData(JsonNode before, JsonNode after, JsonNode source, CdcConnectorMetadata cdcConnectorMetadata) {
+    final ObjectNode base = (ObjectNode) (after.isNull() ? before : after);
```
Comment (davinchia) [inline airbyte-integrations/connectors/source-debezium/src/main/java/io/airbyte/integrations/source/debezium/DebeziumEventUtils.java:64]: if there isn't a good reason to mutate the input argument, we can use `Jsons.clone` to do this

File: airbyte-integrations/connectors/source-debezium/src/main/java/io/airbyte/integrations/source/debezium/FilteredFileDatabaseHistory.java [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+import io.debezium.relational.history.FileDatabaseHistory;
+import io.debezium.relational.history.HistoryRecord;
+import io.debezium.relational.history.HistoryRecord.Fields;
+import io.debezium.relational.history.HistoryRecordComparator;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.function.Consumer;
+
+/**
+ * MySQL Debezium connector monitors the database schema evolution over the time and stores the data
```
Comment (davinchia) [inline airbyte-integrations/connectors/source-debezium/src/main/java/io/airbyte/integrations/source/debezium/FilteredFileDatabaseHistory.java:42]: should we remove 'MYSQL' here?

**Review by davinchia — COMMENTED** [review]
Looks good. Thanks for adding the MYSQL/Postgres PRs. It was only after seeing how this is used that I really understood it.

Because there are alot of files touched, and most I don't think are new, I focused more on the organisation and the interfaces.

1) I think it would make it easier to understand if we moved debezium as a package in the base folder instead of as `source-debezium`. `source-debezium` makes this seem like a source in itself.
2) I think we can do organising the new debezium abstractions. In general the interface and classes meant to be consumed by the public so be top level. My understanding
3) Add comments to the interfaces to make it clear how they are supposed to be used. It was only clear to me after reading the Mysql/Postgres code.

File: airbyte-integrations/connectors/source-debezium/src/testFixtures/java/io/airbyte/integrations/source/debezium/CdcSourceTest.java [hunk 4]
```diff
@@ -0,0 +1,614 @@
...
+        .toListAndClose(secondBatchIterator);
+
+    List<AirbyteStateMessage> stateAfterSecondBatch = extractStateMessages(dataFromSecondBatch);
+    assertEquals(1, stateAfterSecondBatch.size());
+    assertNotNull(stateAfterSecondBatch.get(0).getData());
+    assertExpectedStateMessages(stateAfterSecondBatch);
+
+    Set<AirbyteRecordMessage> recordsFromSecondBatch = extractRecordMessages(
+        dataFromSecondBatch);
+    assertEquals(20, recordsFromSecondBatch.size(),
```
Comment (Phlair) [inline airbyte-integrations/connectors/source-debezium/src/testFixtures/java/io/airbyte/integrations/source/debezium/CdcSourceTest.java:421]: replace 20 with recordsToCreate reference

File: airbyte-integrations/connectors/source-debezium/src/testFixtures/java/io/airbyte/integrations/source/debezium/CdcSourceTest.java [hunk 5]
```diff
@@ -0,0 +1,614 @@
...
+        .read(getConfig(), CONFIGURED_CATALOG, null);
+    final List<AirbyteMessage> dataFromFirstBatch = AutoCloseableIterators
+        .toListAndClose(firstBatchIterator);
+    List<AirbyteStateMessage> stateAfterFirstBatch = extractStateMessages(dataFromFirstBatch);
+    assertEquals(1, stateAfterFirstBatch.size());
+    assertNotNull(stateAfterFirstBatch.get(0).getData());
+    assertExpectedStateMessages(stateAfterFirstBatch);
+    Set<AirbyteRecordMessage> recordsFromFirstBatch = extractRecordMessages(
+        dataFromFirstBatch);
+    assertEquals((MODEL_RECORDS.size() + 20), recordsFromFirstBatch.size());
```
Comment (Phlair) [inline airbyte-integrations/connectors/source-debezium/src/testFixtures/java/io/airbyte/integrations/source/debezium/CdcSourceTest.java:395]: replace 20 with recordsToCreate ref

File: airbyte-integrations/connectors/source-debezium/src/testFixtures/java/io/airbyte/integrations/source/debezium/CdcSourceTest.java [hunk 6]
```diff
@@ -0,0 +1,614 @@
...
+    // first incremental.
+    final Set<AirbyteRecordMessage> recordsFromFirstBatchWithoutDuplicates = removeDuplicates(
+        recordsFromFirstBatch);
+    final Set<AirbyteRecordMessage> recordsFromSecondBatchWithoutDuplicates = removeDuplicates(
+        recordsFromSecondBatch);
+
+    final int recordsCreatedBeforeTestCount = MODEL_RECORDS.size();
+    assertTrue(recordsCreatedBeforeTestCount < recordsFromFirstBatchWithoutDuplicates.size(),
+        "Expected first sync to include records created while the test was running.");
+    assertEquals(40 + recordsCreatedBeforeTestCount,
```
Comment (Phlair) [inline airbyte-integrations/connectors/source-debezium/src/testFixtures/java/io/airbyte/integrations/source/debezium/CdcSourceTest.java:434]: Replace 40 with relevant variable references

**Review by Phlair — COMMENTED** [review]
Looks great, I've made mostly minor comments. Other changes I anticipate being needed are in the abstract cdc test to accommodate MSSQL as we discussed (database->schema->table structure, change in primary key definition on query, order change in creating tables and assigning permissions etc.)

File: airbyte-integrations/bases/debezium/build.gradle [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+    implementation 'io.debezium:debezium-connector-postgres:1.4.2.Final'
+
+    testFixturesImplementation project(':airbyte-db')
+    testFixturesImplementation project(':airbyte-integrations:bases:base-java')
+
+    testFixturesImplementation 'org.junit.jupiter:junit-jupiter-engine:5.4.2'
+    testFixturesImplementation 'org.junit.jupiter:junit-jupiter-api:5.4.2'
+    testFixturesImplementation 'org.junit.jupiter:junit-jupiter-params:5.4.2'
+
+}
```
Comment (davinchia) [inline airbyte-integrations/bases/debezium/build.gradle:23]: nit: newline

**Review by davinchia — APPROVED** [review]

**Review by Phlair — APPROVED** [review]
lgtm!

File: airbyte-integrations/connectors/source-debezium/src/main/java/io/airbyte/integrations/source/debezium/AirbyteSchemaHistoryStorage.java [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+  }
+
+  static AirbyteSchemaHistoryStorage initializeDBHistory(JsonNode schemaHistory) {
+    final Path dbHistoryWorkingDir;
+    try {
+      dbHistoryWorkingDir = Files.createTempDirectory(Path.of("/tmp"), "cdc-db-history");
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+    final Path dbHistoryFilePath = dbHistoryWorkingDir.resolve("dbhistory.dat");
```
[the remark this answers is not in the document]
Comment (subodh1810) [reply airbyte-integrations/connectors/source-debezium/src/main/java/io/airbyte/integrations/source/debezium/AirbyteSchemaHistoryStorage.java]: @Phlair the dbHistoryWorkingDir would be isolated per connector

**Review by sherifnada — APPROVED** [review]
Looks great! Mostly requests for clarification or rename

nit: may seem pedantic but in the future could you add the reading order in the PR description? its super helpful with review

**Pull request merged**